ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1298,9 +1306,205 @@ Set<StoreKey> findMissingKeys(List<StoreKey> keys) throws StoreException {
     return missingKeys;
   }
 
+  /**
+   * Returns the list of keys that are not found in the index from the given input keys. This is the same interface as
+   * {@link #findMissingKeys}, however, the implementation is slightly different. In this method, we leverage the locality
+   * of the input keys to make search faster.
+   *
+   * This method is primarily used in replication thread to find the keys that exist in remote peer, but not in the local
+   * store. There are several unique characteristics about this usage, that would make this method much faster than
+   * {@link #findMissingKeys}
+   * 1. The input keys are basically from the same index segment from remote peer, if they already exist locally, they are
+   *    likely to exist in the same index segment as well.
+   * 2. We don't need to deserialize the entire IndexValue from IndexSegment, we only need to check if the key exists or not.
+   * 3. Input keys are scanned from the beginning of the log in remote peer, so when we try to find they in local store,
+   *    we should do the same, not backwards.
+   *
+   * These three characteristics inspire the new implementation.
+   * 1. This method would keep a cache of several index segment's keys. For each key, we first search it in the cache.
+   * 2. If we can't find the key in the cache, this method would locate the next segment and try to find the key in it.
+   * 3. If we find the key in the next segment, then it will be cached. If not, we fall back to use findKey method. The
+   *    index segment returned by findKey method would then be cached.
+   * 4. We only keep limited amount of index segments in the cache.
+   *
+   * One thing we have to figure out with the cache is when to stop using it. The locality of the input keys only exist
+   * when we already replicated them before but somehow get reset due to compaction, or due to replication from a different
+   * host. We should stop using cache
+   * 1. We stop using cache when we hit unsealed index segment.
+   * 2. We stop using cached when we have more than enough misses from the cache.
+   * @param keys The list of keys that needs to be tested against the index
+   * @return The list of keys that are not found in the index
+   * @throws StoreException
+   */
+  Set<StoreKey> findMissingKeysInBatch(List<StoreKey> keys) throws StoreException {
+    Set<StoreKey> missingKeys = new HashSet<>();
+
+    boolean shouldUseCachedKeys = true;
+    int continuousMissingKeyFromCacheCount = 0;
+
+    // This method should only be invoked in one replication thread, but to make sure there is no race condition, we use
+    // a lock to protect it.
+
+    missingKeysLock.lock();
+    try {
+      for (int i = 0; i < keys.size(); i++) {
+        StoreKey key = keys.get(i);
+        if (!shouldUseCachedKeys) {
+          // if the cache is disabled, then just use a regular findKey method to check the existence of the key
```

Comment:
nit: To avoid this check, may be we could also break from for loop and iterate in another loop when we find cache is disabled.

Reply:
Updated